src/config/datadog.rs

```diff
@@ -0,0 +1,117 @@
+use super::{ComponentId, Config, SinkOuter, SourceOuter};
+use crate::{
+    sinks::datadog::metrics::DatadogConfig, sources::internal_metrics::InternalMetricsConfig,
+};
+use serde::{Deserialize, Serialize};
+
+static INTERNAL_METRICS_KEY: &str = "#datadog_internal_metrics";
+static DATADOG_METRICS_KEY: &str = "#datadog_metrics";
+
+#[derive(Debug, Deserialize, Serialize, PartialEq, Clone)]
+#[serde(default, deny_unknown_fields)]
+pub struct Options {
+    #[serde(default = "default_enabled")]
+    pub enabled: bool,
+
+    #[serde(default = "default_api_key")]
+    pub api_key: Option<String>,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            enabled: default_enabled(),
+            api_key: default_api_key(),
+        }
+    }
+}
+
+/// By default, the Datadog feature is enabled.
+fn default_enabled() -> bool {
+    true
+}
+
+/// By default, no API key is provided.
+fn default_api_key() -> Option<String> {
+    None
+}
+
+/// Augment configuration with observability via Datadog if the feature is enabled and
+/// an API key is provided.
+pub fn attach(config: &mut Config) {
+    // Return early if an API key is missing, or the feature isn't enabled.
+    let api_key = match (&config.datadog.api_key, config.datadog.enabled) {
+        (Some(api_key), true) => api_key.clone(),
+        _ => return,
+    };
+
+    info!("Datadog API key detected. Internal metrics will be sent to Datadog.");
+
+    let internal_metrics_id = ComponentId::from(INTERNAL_METRICS_KEY);
+    let datadog_metrics_id = ComponentId::from(DATADOG_METRICS_KEY);
+
+    // Create an internal metrics source. We're using a distinct source here and not
+    // attempting to reuse an existing one, due to the use of a custom namespace to
+    // satisfy reporting to Datadog.
+    let internal_metrics = InternalMetricsConfig::namespace("pipelines");
```

Comment:
Is this definitely the namespace we want? I didn't see it discussed in the RFC so just want to confirm.

I realize it is pretty easy to change later though.

Reply:
cc @binarylogic 

This is the namespace we'd brought up in Slack / stand-up, but I don't think it's actually documented anywhere.

Worth confirming 👍🏻